### PR TITLE
[ISSUE #5392] [RIP-55] Support remoting protocol in rocketmq proxy module

### LINKED ISSUE #5392: [RIP-55] Support remoting protocol in rocketmq proxy module
## Support remoting protocol in rocketmq proxy module

See [RIP-55](https://github.com/apache/rocketmq/wiki/RIP-55-Support-remoting-protocol-in-rocketmq-proxy-module) for more information.

Task:
- [ ] #5486
- [ ] #5485
- [ ] #5406

### LINKED ISSUE #5406: [RIP-55] [Part-3] [Messaging Implementaion] Support remoting protocol in rocketmq proxy module
See [RIP-55](https://github.com/apache/rocketmq/wiki/RIP-55-Support-remoting-protocol-in-rocketmq-proxy-module) for more information.

### LINKED ISSUE #5485: [RIP-55] [Part-2] [Connection implementation] Client connection management
See [RIP-55](https://github.com/apache/rocketmq/wiki/RIP-55-Support-remoting-protocol-in-rocketmq-proxy-module) for more information.

### LINKED ISSUE #5486: [RIP-55] [Part-1] The remoting server and protocol negotiation

### PR DESCRIPTION
**Pull request by drpmma**
Close #5392
Close #5486
Close #5485
Close #5406

## What is the purpose of the change

Support remoting protocol in the rocketmq proxy module, including the rocketmq-client which version >= 4.9.5 for 4.9.x branch and version >= 5.0.1 for develop branch.

## Brief changelog

See [wiki](https://github.com/apache/rocketmq/wiki/RIP-55-Support-remoting-protocol-in-rocketmq-proxy-module) for more information.

Including three parts.

1. Remoting server implementation and protocol negotiation for both remoting and gRPC protocol.
2. Connection module: Including heartbeat management and synchronization.
3. Message module: Including sending and receiving messages implemented by request to the corresponding broker.

### CODE REVIEW

File: proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/client/ClientActivity.java [hunk 1]
```diff
@@ -413,14 +417,48 @@ protected class ConsumerIdsChangeListenerImpl implements ConsumerIdsChangeListen
...
+        protected void processClientRegister(String group, Object... args) {
+            if (args == null || args.length < 2) {
+                return;
+            }
+            if (args[1] instanceof ClientChannelInfo) {
+                ClientChannelInfo clientChannelInfo = (ClientChannelInfo) args[1];
+                Channel channel = clientChannelInfo.getChannel();
+                if (ChannelHelper.isRemote(channel)) {
+                    // save settings from channel sync from other proxy
+                    Settings settings = GrpcClientChannel.parseChannelExtendAttribute(channel);
```
Comment (ShadowySpirits) [inline proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/client/ClientActivity.java:456]: This looks weird, settings are parsed from the channel but grpcClientSettingsManager is an argument to build Channel. 
@xdkxlk Do we have a cleaner way to resolve this circular dependency?

File: common/src/main/java/org/apache/rocketmq/common/attribute/TopicMessageType.java [hunk 2]
```diff
@@ -40,6 +42,22 @@ public String getValue() {
...
+        } else if (messageProperty.get(MessageConst.PROPERTY_DELAY_TIME_LEVEL) != null
+            || messageProperty.get(MessageConst.PROPERTY_TIMER_DELIVER_MS) != null
+            || messageProperty.get(MessageConst.PROPERTY_TIMER_DELAY_SEC) != null) {
+            return TopicMessageType.DELAY;
+        } else if (messageProperty.get(MessageConst.PROPERTY_SHARDING_KEY) != null) {
+            return TopicMessageType.FIFO;
+        } else {
+            return TopicMessageType.NORMAL;
+        }
+    }
```
Comment (aaron-ai) [inline common/src/main/java/org/apache/rocketmq/common/attribute/TopicMessageType.java:59]: `else` here is redundant seemingly.

File: proxy/src/main/java/org/apache/rocketmq/proxy/common/channel/ChannelHelper.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    public static ChannelProtocolType getChannelProtocolType(Channel channel) {
+        if (channel instanceof GrpcClientChannel) {
+            return ChannelProtocolType.GRPC_V2;
+        } else if (channel instanceof RemotingChannel) {
+            return ChannelProtocolType.REMOTING;
+        } else if (channel instanceof RemoteChannel) {
+            RemoteChannel remoteChannel = (RemoteChannel) channel;
+            return remoteChannel.getType();
+        }
```
Comment (aaron-ai) [inline proxy/src/main/java/org/apache/rocketmq/proxy/common/channel/ChannelHelper.java:46]: ditto.

File: proxy/src/main/java/org/apache/rocketmq/proxy/service/message/ClusterMessageService.java [hunk 4]
```diff
@@ -205,6 +206,32 @@ public CompletableFuture<Long> getMinOffset(ProxyContext ctx, AddressableMessage
...
+    @Override
+    public CompletableFuture<RemotingCommand> request(ProxyContext ctx, String brokerName, RemotingCommand request,
+        long timeoutMillis) {
+        try {
+            String brokerAddress = topicRouteService.getBrokerAddr(brokerName);
+            return mqClientAPIFactory.getClient().invoke(brokerAddress, request, timeoutMillis);
+        } catch (Exception e) {
+            CompletableFuture<RemotingCommand> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
```
Comment (aaron-ai) [inline proxy/src/main/java/org/apache/rocketmq/proxy/service/message/ClusterMessageService.java:218]: use `FutureUtils#completeExceptionally` instead.

File: proxy/src/main/java/org/apache/rocketmq/proxy/service/message/ClusterMessageService.java [hunk 5]
```diff
@@ -205,6 +206,32 @@ public CompletableFuture<Long> getMinOffset(ProxyContext ctx, AddressableMessage
...
+    public CompletableFuture<Void> requestOneway(ProxyContext ctx, String brokerName, RemotingCommand request,
+        long timeoutMillis) {
+        try {
+            String brokerAddress = topicRouteService.getBrokerAddr(brokerName);
+            return mqClientAPIFactory.getClient().invokeOneway(brokerAddress, request, timeoutMillis);
+        } catch (Exception e) {
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+        }
```
Comment (aaron-ai) [inline proxy/src/main/java/org/apache/rocketmq/proxy/service/message/ClusterMessageService.java:232]: ditto.

File: proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/client/ClientActivity.java (hunk 1 above)
Comment (xdkxlk) [reply proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/client/ClientActivity.java:456]: The purpose is different. The parseChannelExtendAttribute method of GrpcClientChannel is a tool method to try to convert the attribute to the Settings. And the grpcClientSettingsManager is passed to channel in order to get extendAttribute of this channel in real time.

**Review by lizhimins — APPROVED** [review]

**Review by aaron-ai — DISMISSED** [review]

**Review by ShadowySpirits — APPROVED** [review]

**Review by lollipopjin — DISMISSED** [review]
LGTM

**Review by hzh0425 — DISMISSED** [review]

File: proxy/src/main/java/org/apache/rocketmq/proxy/service/sysmessage/HeartbeatSyncer.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        if (msgs == null || msgs.isEmpty()) {
+            return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
+        }
+
+        for (MessageExt msg : msgs) {
+            try {
+                HeartbeatSyncerData data = JSON.parseObject(new String(msg.getBody(), StandardCharsets.UTF_8), HeartbeatSyncerData.class);
+                if (data.getConnectProxyIp().equals(ConfigurationManager.getProxyConfig().getLocalServeAddr())) {
+                    continue;
+                }
```
Comment (echooymxq) [inline proxy/src/main/java/org/apache/rocketmq/proxy/service/sysmessage/HeartbeatSyncer.java:188]: If there are two proxy instances on one machine, will there be a problem? The same address cannot distinguish different heartbeat data, maybe we need a unique proxyId?

Comment (drpmma) [reply proxy/src/main/java/org/apache/rocketmq/proxy/service/sysmessage/HeartbeatSyncer.java]: fixed

**Review by lizhimins — APPROVED** [review]

**Pull request merged**